src/mlpack/methods/ann/loss_functions/multilabel_softmargin_loss_impl.hpp

```diff
@@ -0,0 +1,92 @@
+/**
+ * @file multilabel_softmargin_loss_impl.hpp
+ * @author Anjishnu Mukherjee
+ *
+ * Implementation of the Multi Label Soft Margin Loss function.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_IMPL_HPP
+#define MLPACK_METHODS_ANN_LOSS_FUNCTION_MULTILABEL_SOFTMARGIN_LOSS_IMPL_HPP
+
+// In case it hasn't been included.
+#include "multilabel_softmargin_loss.hpp"
+
+namespace mlpack {
+namespace ann /** Artifical Neural Network. */ {
+
+template<typename InputDataType, typename OutputDataType>
+MultiLabelSoftMarginLoss<InputDataType, OutputDataType>::
+MultiLabelSoftMarginLoss(
+    const size_t numClasses,
+    const bool reduction) :
+    numClasses(numClasses),
+    reduction(reduction)
+{
+  classWeights.ones(1, numClasses);
+}
+
+template<typename InputDataType, typename OutputDataType>
+MultiLabelSoftMarginLoss<InputDataType, OutputDataType>::
+MultiLabelSoftMarginLoss(
+    const arma::rowvec& weights,
+    const size_t numClasses,
+    const bool reduction) :
+    numClasses(numClasses),
+    reduction(reduction)
+{
+  classWeights.ones(1, numClasses);
+  classWeights = weights;
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename InputType, typename TargetType>
+typename InputType::elem_type
+MultiLabelSoftMarginLoss<InputDataType, OutputDataType>::Forward(
+    const InputType& input, const TargetType& target)
+{
+  InputType logSigmoid = arma::log((1 / (1 + arma::exp(-input))));
+  InputType logSigmoidNeg = arma::log(1 / (1 + arma::exp(input)));
+  InputType loss = arma::mean(arma::mean(-(target % logSigmoid +
+      (1 - target) % logSigmoidNeg), 1) * classWeights, 1);
```

Comment:
Hey, I am not sure if this is correct. Weights are assigned to each class and input is definitely shuffled, multiplying with weights like this doesn't ensure that we multiplied the first class with weight[0].
Example:
```
groundTruth = [2, 3, 1]
weights = [1, 0.9, 0.5]
loss = reduce([0.9 * loss_function, 0.5 * loss_function, 1 * loss_function])
// Current implementation:
loss = reduce([1 * loss_function, 0.9 * loss_function, 0.5 * loss_function])

Reply:
I have started working on the implementation, I'll open a PR in couple of hours.

Reply:
Can you directly commit to this branch? Or maybe, if you would like, suggest a change in the code review itself using the github feature to do so and I will integrate that in this PR directly?

Reply:
I don't think I can commit directly in your branch. I am making changes in cross-entropy loss function so that it can be used as a reference.

Reply:
> suggest a change in the code review itself using the github feature to do so and I will integrate that in this PR directly?

I am trying to figure out the best way to implement this as well and I don't want to suggest something that doesn't build / work as desired.

Reply:
> input is definitely shuffled
Is this mentioned somewhere? 
The [PyTorch](https://github.com/pytorch/pytorch/blob/a91097bdfb0618e15bb5b7ec96869da3c7ba278a/torch/nn/functional.py#L2666) implementation of the function does not have shuffling involved as far as I can see.

Reply:
> I don't think I can commit directly in your branch. 

Look at point 6 of [this](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/commenting-on-a-pull-request) link. Can you do that for this PR?

Reply:
Ohk, so here is weighted loss from documentation : `\text{loss}(x, class) = weight[class] \left(-x[class] + \log\left(\sum_j \exp(x[j])\right)\right)`
As you can see loss of each class is multiplied with its weight (weight[class]). The reason we do this is : The input might not be of the format 0, 1, 2 and hence we multiply loss as weight[0], weight[1] and weight[2].  I think you can refer to weighted loss for more info. 

I can try that as well.

Reply:
> I think you can refer to weighted loss for more info.

Could you post the link for this please? I can't seem to be able to find it🙈

Reply:
weights = {1: 1
		   2: 0.9
		   3: 0.5};

grounTruth = {3, 2, 1};
input = {2.5, 1.5, 0.5};
loss_function = abs(input - target);
loss = reduction({0.5, 0.5, 0.5});
// if reduction == sum => loss = sum(loss) = 1.5;
loss_function = reduce({0.5 * 0.5, 0.9 * 0.5, 1 * 0.5});

@ojhalakshya will explain it better.

Reply:
Sure! 

> Hey, I am not sure if this is correct. Weights are assigned to each class and input is definitely shuffled, multiplying with weights like this doesn't ensure that we multiplied the first class with weight[0].
> Example:
> 
> ```
> groundTruth = [2, 3, 1]
> weights = [1, 0.9, 0.5]
> loss = reduce([0.9 * loss_function, 0.5 * loss_function, 1 * loss_function])
> // Current implementation:
> loss = reduce([1 * loss_function, 0.9 * loss_function, 0.5 * loss_function])
> ```

As far as I can see for now,
```
  rowvec A = arma::mat("1 0.9 0.5");
  rowvec B = shift(A, -1);
```

should achieve this shuffle in the weight vector that you have given an example of. 
So, are you suggesting I do this to the weights vector, before it is multiplied?
@kartikdutt18

Reply:
Hi @iamshnoo 
I am attaching a photo of some explanations that might help here based on the previous conversations above, I have made them relatable.
Do tell me if it helps.
![WhatsApp Image 2020-04-10 at 10 20 39 PM](https://user-images.githubusercontent.com/57477999/79008013-1a8e9580-7b7a-11ea-8f6d-87531c47bb95.jpeg)

Thanks.

Reply:
> ```
> groundTruth = [2, 3, 1]
> ```

The groundTruth or the class labels will be one-hot encoded. See the test case for an example.

So, basically
```
target = [ [0,1,0], [0,0,1], [1,0,0] ]
```

> The input might not be of the format 0, 1, 2 

So, I don't understand what you mean by this. Maybe some more explanation about this would be helpful and also the source from where you are referring this. 

> hence we multiply loss as weight[0], weight[1] and weight[2].

When i write a = b * w (the current implementation),
isn't that basically saying a = [ b[0] * w[0], b[1] * w[1], b[2] * w[2] ]. So, what exactly are you asking to change? I am not quite sure about this either I guess.

Reply:
> I am attaching a photo of some explanations that might help here based on the previous conversations above, I have made them relatable.

I am taking a look at this and will get back in a few minutes

Reply:
Hmm, Do you mind running tests with weight matrices, shuffled input and verifying from pytorch?

Reply:
Hi @kartikdutt18  @ojhalakshya can you please take a look at this. 
![explanation](https://user-images.githubusercontent.com/45921510/79011230-3b0e1e00-7b81-11ea-801d-e154c2ebe354.jpeg)

Thanks!

Reply:
> Hmm, Do you mind running tests with weight matrices, shuffled input and verifying from pytorch?

Sure I can. Can you pleasehiave a look at the image I posted above and mention the inputs, targets and weights you want me to test with?

Reply:
I think you can decide that, I am fine with anything as long as it ensures that this implementation does what it intends to do.

Reply:
Cool! But, do you realize the point I am trying to make in the image I posted?

Reply:
Yes I do, unfortunately I can't give a lot of time to this PR or the one that I just opened. I am working on another PR that is kind of top-priority for me. If I get a chance, I will get back to you tomorrow with a working code in the other PR. Also I will provide a reference for the same. I look forward to your test results (with the current implementation).

Reply:
Okay! Thank you so much for all the help :)

Reply:
Ohk, I was going through the some references, this is what I found:
1. This implementation only works for OHE.
2. Multi-Labels loss functions will always have input as OHE.

This will fail if that isn't the case. So maybe this will work (after you add tests).
What I was talking about was the case when the target wasn't OHE as that is the case with many loss functions.

Reply:
Reference: https://arxiv.org/pdf/1702.05659.pdf

Reply:
> 1. This implementation only works for OHE.
> 2. Multi-Labels loss functions will always have input as OHE.

Exactly my point too. Thank you for verifying this.
I will definitely write down some testing for more cases as we discussed today and get back to you tomorrow with this.

Reply:
@kartikdutt18 thanks for the excellent inputs. There is a lot going on in this thread, I don't think I am all caught up with it currently. Maybe I'll take some time and then review or suggest if I think any further development is needed.
@iamshnoo great work!
@ojhalakshya thanks for taking out time to write it down. It is highly appreciated. But I think what you have written makes sense for a regression problem. I am not sure if it works in this particular scenario. Maybe I am missing something or I need to take a closer look :)

However, I think I will just mention this once as a reminder (although I am sure everyone knows this): The output of the last layer of nodes of a neural network (called **input** in mlpack) is fed to the loss function. So for a classification task, there **has** to be one node for each class. Instead of *n_classes* number of output nodes if just one node is there, then it will technically be a regression task (in which case the loss function should be differently written).